### ISSUE TITLE
Add code formatting tool

### ISSUE THREAD

**Issue by jooola** [airsonic/airsonic#1391]
I would be nice to add a code formatting tool in this projectn for IDE's and CI.

The question is which on should we use.
Here are some suggestion:
https://github.com/google/google-java-format
https://github.com/diffplug/spotless
https://github.com/jhipster/prettier-java

I would go for Google plugin if we want to start a code formatting dictatorship. But I'm good with this.

Prettier use nodejs, so this might not be a great idea.

**Comment by jvoisin** [comment]
We already have  checkstyle #1121

**Comment by jooola** [comment]
Does checkstyle format the files ? I mean it seems more like a linting tool, not a formatting tool.
I purely speaking code readability. Both could be combined though ?

**Comment by jvoisin** [comment]
Both should be combined, yes :)

**Comment by tesshucom** [comment]
Checkstyle and formatter are the foundations of joint development.
The most popular formatter-maven-plugin doesn't seem to be listed, but anything that has equivalent functionality is fine.

 - Generate configuration xml using Eclipse settings GUI.
 - We can use the xml with Maven
 - We can use the xml with IntelliJ

**Comment by randomnicode** [comment]
Hmmm, what would the workflow be?

Who's going to commit the changes?

If travis is doing the formatting and committing changes during a build, then we need to get travis some creds. Additionally that runs the risk of causing some merge conflicts especially on stacked PRs. Consider:
- You build branch1 with changes to File 1
- You build branch2 on top of branch1 with more changes to File 1
- You submit both as stacked PRs
- You merge PR1 (on branch1) to master
- Travis does a formatting commit on master for File 1
- Now due to the formatting commit being on the same file, PR2 (on branch 2) has merge conflicts even though it was stacked on top of PR1 and there were no changes made by anyone else in the middle.

If the devs are going to do it prior to a commit, then we don't need an external formatter plugin. About every IDE comes with a formatter. Just need to import a formatting guideline export into the IDE.

I'm not against the idea, maybe I just don't understand it?

Also, just my opinion, but I do find our checkstyle to be too restrictive.

**Comment by tesshucom** [comment]
We can modify thousands of code by simply testing existing features and fixing problems.
Even if someone does it locally, those who can predict what will happen in the review do not want to provide feedback upstream.

CheckStyle, Formatter, Findbugs and PMD are common enough tools that has existed since ancient times.
There are many tiny probrems that have been pointed out in reviews, that can be detected by tools.
Some people think this responsibility lies with the committer, while others think it is in the project management.

However, that seems to have a big gap in person.
If the maintainer feels it is painful to control them, I cannot strongly recommend it.

**Comment by randomnicode** [comment]
These are all static analysis tools. About none of them do any kind of commits. They can fail builds, but very rarely have I seen them actually commit on behalf of any committers to _correct_ a failing build automatically which is what at least I interpreted the original issue to be. The impetus to following a standard convention would have to come from a developer locally, and you're correct in mentioning we could check in an xml or some kind of formatting artifact to help the IDE follow a standard, but I really don't see a need for a plugin.

Additionally, while I totally agree that there should be a standard to formatting code, please note that formatting differently _itself_ is not a bug. It might breed them due to poor readability, but that's about it. Checkstyle may fail a build due to an indent level gone wrong, or import statements in different order, but none of these are bugs, just convention. They're all still part of the Java Language spec, and the compiler won't see them any differently. Formatters don't fix bugs, they merely prettify and provide a means to coalesce around and adhere to a human readable convention (similar to how a minifier doesn't change the semantics of a JS file).

Also please note that FindBugs has now effectively been succeeded by SpotBugs.

**Comment by tesshucom** [comment]
> Also, just my opinion, but I do find our checkstyle to be too restrictive.

I don't think so severe.
However, since there is a flaw in the way, I agree if it puts a burden on the committer.

Usually check style and formatter are applied at the same time.
If only the check style is applied manually, only the conformity to the check style is imposed on the human power of the committer.
This is the current state.

The idea of ​​distributing an IDE formatter is also indicated.
If you agree with that idea, it is probably not much different from the direction that this issue's title shows.

The formatter definition of eclipse and maven close to the median of IDE and development environment can be shared.
IntelliJ can import Eclipse format.

In other words, creating a generic code formatting tool definition is close to providing coding controls and automatic formatting in the IDE.

> Additionally, while I totally agree that there should be a standard to formatting code, please note that formatting differently itself is not a bug.

I agree with that opinion.
It ’s better to spend more time on adding tests.

However, I know from past experience that Airsonic maintainers love to stick with formatting review.
Because these are not efficient, I think that these people should do the tool maintenance quickly.

**Comment by tesshucom** [comment]
> Also, just my opinion, but I do find our checkstyle to be too restrictive.

I don't think so severe.
However, since there is a flaw in the way, I agree if it puts a burden on the committer.

Usually check style and formatter are applied at the same time.
If only the check style is applied manually, only the conformity to the check style is imposed on the human power of the committer.
This is the current state.

The idea of ​​distributing an IDE formatter is also indicated.
If you agree with that idea, it is probably not much different from the direction that this issue's title shows.

The formatter definition of eclipse and maven close to the median of IDE and development environment can be shared.
IntelliJ can import Eclipse format.

In other words, creating a generic code formatting tool definition is close to providing coding controls and automatic formatting in the IDE.

> Additionally, while I totally agree that there should be a standard to formatting code, please note that formatting differently itself is not a bug.

I agree with that opinion.
It ’s better to spend more time on adding tests.

However, I know from past experience that maintainers love to stick with formatting review.
Tool maintenance pushes this load down.

**Comment by tesshucom** [comment]
> Also, just my opinion, but I do find our checkstyle to be too restrictive.

I don't think so severe.
However, since there is a flaw in the way, I agree if it puts a burden on the committer.

Usually check style and formatter are applied at the same time.
If only the check style is applied, only the conformity to the check style is imposed on the human power of the committer.
This is the current state.

The idea of ​​distributing an IDE formatter is also indicated.
If you agree with that idea, it is probably not much different from the direction that this issue's title shows.

The formatter definition of eclipse and maven close to the median of IDE and development environment can be shared.
IntelliJ can import Eclipse format.

In other words, creating a generic code formatting tool definition is close to providing coding controls and automatic formatting in the IDE.

> Additionally, while I totally agree that there should be a standard to formatting code, please note that formatting differently itself is not a bug.

I agree with that opinion.
It ’s better to spend more time on adding tests.

However, I know from past experience that maintainers love to stick with formatting review.
Tool maintenance pushes this load down.
Emotions and inhomogeneities are eliminated and rules can be controlled.

**Comment by fxthomas** [comment]
We need something like checkstyle, precisely for the reason @tesshucom evoked, that discussions on code formatting are pushed from individual issues towards specific issues about style rules.

I'm neutral on having a formatting tool. If anybody wants to merge one then let's do it, especially if it can read checkstyle rules.

**Comment by tesshucom** [comment]
OK.
I don't want to sound as if I'm giving orders.
I realized that these were not perceived as serious problems by key members.
Thank you very much.
